src/metrics/include/metrics/v2/event.hpp

```diff
@@ -0,0 +1,279 @@
+            return;
+        }
+        if (!_stream->WritesDone()) {
+            BOOST_LOG_TRIVIAL(warning) << "Closing grpc stream, but not all writes completed.";
+        }
+        auto status = _stream->Finish();
+        if (!status.ok()) {
+            BOOST_LOG_TRIVIAL(error)
+                << "Problem closing grpc stream: " << _context.debug_error_string();
+        }
+    }
+
+private:
+    grpc::WriteOptions _options;
+    poplar::PoplarResponse _response;
+    grpc::ClientContext _context;
+    std::unique_ptr<grpc::ClientWriterInterface<poplar::EventMetrics>> _stream;
+};
+
+
+/**
+ * Manages the gRPC-side collector for each operation.
+ * Exists for construction / destruction resource management only.
+ */
+class Collector {
+public:
+    Collector(const Collector&) = delete;
+
+    explicit Collector(CollectorStubInterface& stub,
+                       std::string name,
+                       const std::string& path_prefix)
+        : _name{std::move(name)}, _stub{stub}, _id{} {
+        _id.set_name(_name);
+
+        grpc::ClientContext context;
+        poplar::PoplarResponse response;
+        poplar::CreateOptions options = createOptions(_name, path_prefix);
+        auto status = _stub->CreateCollector(&context, options, &response);
+
+        if (!status.ok()) {
+            std::ostringstream os;
+            os << "Collector status not okay: " << status.error_message();
+            BOOST_THROW_EXCEPTION(PoplarRequestError(os.str()));
+        }
+    }
+
+    ~Collector() {
+        grpc::ClientContext context;
+        poplar::PoplarResponse response;
+        auto status = _stub->CloseCollector(&context, _id, &response);
+        if (!status.ok()) {
+            BOOST_LOG_TRIVIAL(error) << "Couldn't close collector: " << status.error_message();
+        }
+    }
+
+    CollectorStubInterface& _stub;
+
+private:
+    static auto createPath(const std::string& name, const std::string& path_prefix) {
+        std::stringstream str;
+        str << path_prefix << '/';
+        str << name << ".ftdc";
+        return str.str();
+    }
+
+    static poplar::CreateOptions createOptions(const std::string& name,
+                                               const std::string& path_prefix) {
+        poplar::CreateOptions options;
+        options.set_name(name);
+        options.set_events(poplar::CreateOptions_EventsCollectorType_BASIC);
+        options.set_path(createPath(name, path_prefix));
+        options.set_chunksize(1000);
+        options.set_streaming(true);
+        options.set_dynamic(false);
+        options.set_recorder(poplar::CreateOptions_RecorderType_PERF);
+        options.set_events(poplar::CreateOptions_EventsCollectorType_BASIC);
+        return options;
+    }
+
+    std::string _name;
+    poplar::PoplarID _id;
+};
+
+/**
+ * Class that keeps track of the duration and totals for an operation across all threads.
+ */
+template <typename ClockSource>
+struct DurationCounter {
```

Comment:
This doesn't actually track duration across all threads, just for the current thread. I think I'll cut another ticket to make it actually track across all threads because I'm not sure of the best way to architecture that yet.

Reply:
We shouldn't ever need cross-thread duration since the grpc client is supposed to do that for us (albeit it has to lock because...go...)

Reply:
Slightly confused what you mean by that, this is for the `duration` field in the `EventMetrics` message, which we need to set manually? Should probably chat in-person about that.

Reply:
After an in-person discussion, we learned that the cross-thread summation can be done automatically poplar-side if we have cross-thread collectors.